packages/contracts/contracts/ChugSplashRegistry.sol

```diff
@@ -1,50 +1,312 @@
+    }
+
+    /**
+     * Allows a manager to approve a bundle to be executed. Note that the bundle can be executed
+     * as soon as the bundle is approved.
+     *
+     * @param _name Name of the ChugSplash project.
+     * @param _bundleHash Hash of the bundle to approve.
+     */
+    function approveChugSplashBundle(string memory _name, bytes32 _bundleHash) public onlyManager(_name) {
+        ChugSplashProject storage project = projects[_name];
+        require(
+            project.status == ChugSplashBundleStatus.PROPOSED,
+            "ChugSplashRegistry: bundle either does not exist or has already been approved or completed"
+        );
+
+        require(
+            isUpgrading(_name) == false,
+            "ChugSplashRegistry: another bundle has been approved and not yet completed"
+        );
+
+        project.activeBundleHash = _bundleHash;
+        project.status = ChugSplashBundleStatus.APPROVED;
+    }
+
+    /**
+     * Executes a specific action within the current active bundle for a project. Actions can only be
+     * executed once. If executing this action would complete the bundle, will mark the bundle as
+     * completed and make it possible for a new bundle to be approved.
+     *
+     * @param _name Name of the ChugSplash project.
+     * @param _action Action to execute.
+     * @param _actionIndex Index of the action in the bundle.
+     * @param _proof Merkle proof of the action within the bundle.
+     */
+    function executeChugSplashBundleAction(
+        string memory _name,
+        ChugSplashAction memory _action,
+        uint256 _actionIndex,
+        bytes32[] memory _proof
+    ) public {
+        require(
+            isUpgrading(_name) == true,
+            "ChugSplashRegistry: no bundle has been approved for execution"
+        );
+
+        ChugSplashProject storage project = projects[_name];
+
+        // TODO: Confirm that this will do out-of-bounds checks.
+        require(
+            project.executions[_actionIndex] == false,
+            "ChugSplashRegistry: action has already been executed"
+        );
+
+        require(
+            MerkleTree.verify(
+                project.activeBundleHash,
+                keccak256(abi.encode(_action.actionType, _name, _action.data)),
+                _actionIndex,
+                _proof,
+                project.executions.length
+            ),
+            "ChugSplashRegistry: invalid bundle action proof"
+        );
+
+        // Make sure the proxy has code in it and deploy the proxy if it doesn't. Since we're
+        // deploying via CREATE2, we can always correctly predict what the proxy address *should*
+        // be and can therefore easily check if it's already populated.
+        // TODO: See if there's a better way to handle this case because it messes with the gas
+        // cost of SET_CODE/SET_STORAGE operations in a somewhat unpredictable way.
+        ChugSplashProxy proxy = getProxyByName(_name);
```

Comment:
Not really related to this PR, but I realized a few weeks ago that it'd probably be better to support standard proxies (or even custom proxies that end-users provide) instead of forcing people to use the `ChugSplashProxy`. One way to do this is to upgrade the contract to a temporary implementation that has the `setStorage` function so we can basically treat any proxy as a "ChugSplashProxy" even if it isn't really one. Anyway, we can talk about that separately, just was reminded about it while reading this.